### DISCUSSION TITLE
Extending the User Model in Prisma with EmailProvider

### DISCUSSION THREAD

**Discussion by VaZark** [nextauthjs/next-auth#4438] [Help]
### Issue

Unable to view additional fields in User Model. 

### Setup 

1. PrismaAdapter
2. Email Provider (updated the Provider as per [docs](https://next-auth.js.org/configuration/providers/oauth#override-default-options)
3. JWT
4. Uses Middleware (as per these [docs](https://next-auth.js.org/configuration/nextjs#default-withauth-export)

### Prisma Model

```prisma

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  role          String    @default("Agent")
  accounts      Account[]
  sessions      Session[]
}
```

### Adapter Config

```ts

export default NextAuth({
  adapter: PrismaAdapter(prisma),
  providers: [
    EmailProvider({
      server: {
        host: process.env.EMAIL_SERVER_HOST,
        port: process.env.EMAIL_SERVER_PORT,
        maxAge: 24 * 60 * 60, // How long email links are valid for (default 24h)
        auth: {
          user: process.env.EMAIL_SERVER_USER,
          pass: process.env.EMAIL_SERVER_PASSWORD
        }
      },
      from: process.env.EMAIL_FROM,
      profile(profile) {
        return {
          id: profile.sub,
          name: profile.name,
          email: profile.email,
          image: profile.image,
          role: profile.role,
        }
      }
    }),
  ],
  session: {
    strategy: "jwt",
    maxAge: 5 * 24 * 60 * 60,
    updateAge: 12 * 60 * 60, // 12 hours
  },
  jwt: {
    maxAge: 60 * 60 * 24 * 5,
  },
  callbacks: {
    async jwt({ token, user, account, profile, isNewUser }) {
      console.log("jwt", token);
      return token
    },
    async session({ session, token }) {
      console.log("token", token);
      console.log("session", session);
      // console.log("user", user);
      return session
    }
  }
})
```
### Log

```

jwt {
  name: 'Vaz',
  email: 'vmohan@lenbox.io',
  picture: null,
  sub: 'cl291qldx0006gizqyx55o29a',
  iat: 1650555761,
  exp: 1650987761,
  jti: '3f718c81-3604-4294-bfc7-06834d23c6bb'
}
token {
  name: 'Vaz',
  email: 'vmohan@lenbox.io',
  picture: null,
  sub: 'cl291qldx0006gizqyx55o29a',
  iat: 1650555761,
  exp: 1650987761,
  jti: '3f718c81-3604-4294-bfc7-06834d23c6bb'
}
session {
  user: { name: 'Vaz', email: 'vmohan@lenbox.io', image: null },
  expires: '2022-04-26T15:43:02.299Z'
}
```

**Comment by arenddeboer** [comment]
According to the TypeScript typings, profile is not part of the EmailProvider provider:
https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/providers/email.ts

Your best bet for adding roles is following this advice: https://github.com/nextauthjs/next-auth/discussions/805#discussioncomment-2618303
The only issue is that session.user is not typed as the Prisma User model, but some nextauth default provider interface.
Haven't figured that part out yet.

**Comment by axelgar** [comment]
Hey @VaZark not sure if this will solve the issue for you but in my case and after trying to understand how the Prisma adapter works this is what I implemented: 

``` ts
export const authOptions: NextAuthOptions = {
  adapter: {
    ...PrismaAdapter(prisma),
    updateUser: ({ id, ...data }) => prisma.user.update({ where: { id }, data, include: { studios: true } }),
  },
  providers: [
    EmailProvider({
      server: process.env.EMAIL_SERVER,
      from: process.env.EMAIL_FROM,
    }),
  ],
  session: {
    strategy: 'jwt',
  },
  callbacks: {
      async jwt({ token, user, ...rest }) {
       // The first time this runs user has the extended field in my case studios
        if (user) {
          token.user = {
            id: user.id,
            role: user.role,
          };
        }
        return token;
      },
    },
  // more code
};
```

Basically, in my case, I had to extend the user with a `studios` relational field and I found out, after trying with `getUser`, `getUserByEmail`, `getUserByAccount`, and `getSessionAndUser` from the [Prisma adapter](https://github.com/nextauthjs/next-auth/blob/main/packages/adapter-prisma/src/index.ts#L221) that it was the`updateUser` method the one that ends up returning the user for the jwt token, in this specific case at least, so I overwrote the implementation to return the user with extended data. In the code example, it's missing to add the extended data to the `token` so that can be retrieved in the `session` callback.

I hope this helps 🙂
